service/header/store.go

```diff
@@ -0,0 +1,302 @@
+package header
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+
+	"github.com/celestiaorg/celestia-core/libs/bytes"
+)
+
+// TODO(@Wondertan): Those values must be configurable and proper defaults should be set for specific node type.
+const (
+	// DefaultStoreCache defines the amount of max entries allowed in the Header Store cache.
+	DefaultStoreCache = 1024
+	// DefaultIndexCache defines the amount of max entries allowed in the Height to Hash index cache.
+	DefaultIndexCache = 256
+)
+
+var (
+	storePrefix = datastore.NewKey("headers")
+	headKey     = datastore.NewKey("head")
+)
+
+type store struct {
+	ds    datastore.Batching
+	cache *lru.ARCCache
+	index *heightIndexer
+
+	headLk sync.RWMutex
+	head   bytes.HexBytes
+}
+
+func NewStore(ds datastore.Batching) (Store, error) {
+	return newStore(ds)
+}
+
+func NewStoreWithHead(ds datastore.Batching, head *ExtendedHeader) (Store, error) {
+	store, err := newStore(ds)
+	if err != nil {
+		return nil, err
+	}
+
+	err = store.put(head)
+	if err != nil {
+		return nil, err
+	}
+
+	return store, store.newHead(head.Hash())
+}
+
+func newStore(ds datastore.Batching) (*store, error) {
+	ds = namespace.Wrap(ds, storePrefix)
+	cache, err := lru.NewARC(DefaultStoreCache)
+	if err != nil {
+		return nil, err
+	}
+
+	index, err := newHeightIndexer(ds)
+	if err != nil {
+		return nil, err
+	}
+
+	return &store{
+		ds:    ds,
+		cache: cache,
+		index: index,
+	}, nil
+}
+
+func (s *store) Start(context.Context) error {
+	err := s.loadHead()
+	if err == datastore.ErrNotFound {
+		return fmt.Errorf("header/store: no chain head")
+	}
```

Comment:
Won't this always happen if the first constructor is used to init this?
```
func NewStore(ds datastore.Batching) (Store, error) {
	return newStore(ds)
}
```

Reply:
This is why for the first time you have to use `NewStoreWithHead` and then `NewStore`. Not sure thought how convenient this is.

Reply:
This is not obvious at all and this is not documented anywhere IIUC.

Reply:
I was confused about this msg, as I remember explaining this with comments, but later realized I didn't commit that part 😅